src/comm.cc

```diff
@@ -1579,6 +1579,12 @@ checkTimeouts(void)
 void
 commStartHalfClosedMonitor(int fd)
 {
+    if (!TheHalfClosed->empty()) {
+        debugs(5, 5, HERE << "half_closed is on fd = " << fd );
+        commCallCloseHandlers(fd);
```

Comment:
commStartHalfClosedMonitor() should not Call Close Handlers. Close handlers should be called on read errors. And they should remain set until the connection is actually being closed. I know this change stops this assertion from happening, but so does disabling the obviously broken `half_closed_client` feature. The change is trading a loud assertion for silent uncertainty. This is not a good tradeoff IMO.

I am suggesting a path towards a possibly acceptable workaround in a dedicated change request.